File: packages/engine/src/cpus/6502/instructions/plp.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('plp', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.sp = 0xff
  cpu.zero = true
  mmu.data = [0x08, 0x28] // PHP PLP
  const g = cpu.run()
  for (let i = 0; i < 3; i++) g.next()
  expect(cpu.sp).toBe(0xfe)
  expect(mmu.data[0x1ff]).toBe(cpu.status | 0x10)

  cpu.zero = false
  for (let i = 0; i < 4; i++) g.next()
  expect(cpu.sp).toBe(0xff)
  expect(cpu.breakCommand).toBeFalsy()
  expect(cpu.zero).toBeTruthy()
})

File: packages/engine/src/cores/famibob/cpu/instructions/plp.test.zig
const testing = @import("../testing/setup.zig");

test "PLP" {
    var cpu = testing.createCPU();
    cpu.sp = 0xff;
    cpu.z = true;
    cpu.write(0, 0x8);
    cpu.write(1, 0x28);
    for (0..3) |_| {
        cpu.process();
    }
    try testing.expectEqual(0xfe, cpu.sp);
    try testing.expectEqual(cpu.getP() | 0x10, cpu.read(0x1ff));

    cpu.z = false;
    for (0..4) |_| {
        cpu.process();
    }
    try testing.expectEqual(0xff, cpu.sp);
    try testing.expectEqual(false, cpu.b);
    try testing.expectEqual(true, cpu.z);
}
